src/main/java/iterator/RichIterator.java

```diff
@@ -477,7 +905,33 @@ default RichIterator<A> distinct() {
      * @return the iterator
      */
     default <B> RichIterator<B> flatMap(Function<? super A, ? extends Iterator<B>> f) {
-        throw new NotImplementedException();
+        return new RichIterator<B>() {
+            private Iterator<B> iterator;
+
+            @Override
+            public boolean hasNext() {
+                if (iterator != null && iterator.hasNext()) {
+                    return true;
+                } else if (RichIterator.this.hasNext()) {
+                    iterator = f.apply(RichIterator.this.next());
+
+                    return iterator.hasNext() || hasNext();
```

Comment:
The recursive code is not very efficient (not critical) because it accumulates stack trace unnecessarily.
Writing an iterative code in this case is more efficient.

something like:
```java
while(!iterator.hasNext() && RichIterator.hasNext()) {
  iterator = apply(RichIterator.this.next());
}
return iterator.hasNext();
```